src/opaque-types-type-alias-impl-trait.md

```diff
@@ -0,0 +1,62 @@
+# Opaque types (type alias `impl Trait`)
+
+Opaque types are syntax to declare an opaque type alias that only
+exposes a specific set of traits as their interface; the concrete type in the
+background is inferred from a certain set of use sites of the opaque type.
+
+This is expressed by using `impl Trait` within type aliases, for example:
+
+```rust,ignore
+type Foo = impl Bar;
+```
+
+This declares an opaque type named `Foo`, of which the only information is that
+it implements `Bar`. Therefore, any of `Bar`'s interface can be used on a `Foo`,
+but nothing else (regardless of whether it implements any other traits).
+
+Since there needs to be a concrete background type, you can currently
+express that type by using the opaque type in a "defining use site".
+
+```rust,ignore
+struct Struct;
```

Comment:
Shouldn't it be `struct Foo;` ?

Reply:
Yes it should! Feel free to open a PR with a fix :)

Reply:
Then the whole example should be

```rust
fn foo() -> impl Bar {
    struct Foo;
    impl Bar for Foo { /* stuff */ }
    Foo
}
```

?

Reply:
Sorry, I misread earlier. Really, this code snippet should be read in the context of the previous one. But that's confusing, so maybe we should add the definition of `Foo` here too. So the complete snippet would be:
```rust
type Foo = impl Bar;

struct Struct;

impl Bar for Struct { /* stuff */ }

fn foo() -> Foo {
    Struct
}
```